storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java

```diff
@@ -0,0 +1,178 @@
+            }
+            return td;
+        }
+
+        /**
+         * Get a score for the simulated user.  This is used to sort the users, by their highest priority topology.
+         * Only give user guarantees that will not exceed cluster capacity.
+         * Score of each resource type is calculated as: (Requested + Assigned - Guaranteed)/clusterAvailable
+         * The final score is a max over all resource types.
+         * Topology score will fall into the following intervals if:
+         *      User is under quota (guarantee):                    [(-guarantee)/available : 0]
+         *      User is over quota:                                 (0, infinity)
+         * Unfortunately, score below 0 does not guarantee that the topology will be scheduled due to resources fragmentation.
+         * @param availableCpu available CPU on the cluster.
+         * @param availableMemory available memory on the cluster.
+         * @param availableGenericResources available generic resources (other that cpu and memory) in cluster
+         * @param td the topology we are looking at.
+         * @return the score.
+         */
+        protected double getScore(double availableCpu, double availableMemory,
+                                  Map<String, Double> availableGenericResources, TopologyDetails td) {
+            // calculate scores for cpu and memory first
+            double ret = super.getScore(availableCpu, availableMemory, td);
+            if (ret == Double.MAX_VALUE) {
+                return ret;
+            }
+            Map<String, Double> tdTotalRequestedGeneric = td.getTotalRequestedGenericResources();
+            if (tdTotalRequestedGeneric == null) {
+                tdTotalRequestedGeneric = Collections.emptyMap();
+            }
+            for (Map.Entry<String, Double> entry : availableGenericResources.entrySet()) {
+                String resource = entry.getKey();
+                Double available = entry.getValue();
+                if (available <= 0) {
+                    return Double.MAX_VALUE;
+                }
+                Double wouldBeResource = assignedGenericResources.getOrDefault(resource, 0.0)
+                    + tdTotalRequestedGeneric.getOrDefault(resource, 0.0);
+                double thisScore = (wouldBeResource -  guaranteedGenericResources.getOrDefault(resource, 0.0)) / available;
+                ret = Math.max(ret, thisScore);
+            }
+
+            return ret;
+        }
+
+        protected double getScore(double availableCpu, double availableMemory, Map<String, Double> availableGenericResources) {
+            TopologyDetails td = getNextHighest();
+            return getScore(availableCpu, availableMemory, availableGenericResources, td);
+        }
+    }
+
+    private static class GrasSimulatedUserComparator implements Comparator<GrasSimulatedUser> {
```

Comment:
Is this comparator supposed to compare GrasSimulatedUser for use in a sort? Then it should compare o1 to o2. But it "seems" to be comparing  "distance" from some fixed point to o1 and o2. Is this intentional?

Reply:
It is comparing GrasSimulatedUser  for sorting at line 58. And it is comparing o1 and o2 based on their scores, please refer to the overriding compare method at line 175.